sympy/matrices/matrices.py

```diff
@@ -2044,6 +2192,425 @@ def minor_submatrix(self, i, j):
+        speeds things up.
+
+        returns a tuple (mat, perm) where perm is a permutation
+        of the columns to perform to shift the complex columns righ, and mat
+        is the permuted matrix."""
+
+        def complexity(i):
+            # the complexity of a column will be judged by how many
+            # element's zero-ness cannot be determined
+            return sum(1 if iszerofunc(e) is None else 0 for e in self[:, i])
+        complex = [(complexity(i), i) for i in range(self.cols)]
+        perm = [j for (i,j) in sorted(complex)]
+
+        return (self.permute(perm, orientation='cols'), perm)
+
+    def _row_reduce(self, iszerofunc, simpfunc, normalize_last=True,
+                    normalize=True, zero_above=True):
+        """Row reduce `self` and return a tuple (rref_matrix,
+        pivot_cols, swaps) where pivot_cols are the pivot columns
+        and swaps are any row swaps that were used in the process
+        of row reduction.
+
+        Parameters
+        ==========
+
+        iszerofunc : determines if an entry can be used as a pivot
+        simpfunc : used to simplify elements and test if they are
+            zero if `iszerofunc` returns `None`
+        normalize_last : indicates where all row reduction should
+            happen in a fraction-free manner and then the rows are
+            normalized (so that the pivots are 1), or whether
+            rows should be normalized along the way (like the naive
+            row reduction algorithm)
+        normalize : whether pivot rows should be normalized so that
+            the pivot value is 1
+        zero_above : whether entries above the pivot should be zeroed.
+            If `zero_above=False`, an echelon matrix will be returned.
+        """
+        rows, cols = self.rows, self.cols
+        mat = list(self)
+        def get_col(i):
+            return mat[i::cols]
+
+        def row_swap(i, j):
+            mat[i*cols:(i + 1)*cols], mat[j*cols:(j + 1)*cols] = \
+                mat[j*cols:(j + 1)*cols], mat[i*cols:(i + 1)*cols]
+
+        def cross_cancel(a, i, b, j):
+            """Does the row op row[i] = a*row[i] - b*row[j]"""
+            q = (j - i)*cols
+            for p in range(i*cols, (i + 1)*cols):
+                mat[p] = a*mat[p] - b*mat[p + q]
+
+        piv_row, piv_col = 0, 0
+        pivot_cols = []
+        swaps = []
+        # use a fraction free method to zero above and below each pivot
+        while piv_col < cols and piv_row < rows:
+            pivot_offset, pivot_val, \
+            assumed_nonzero, newly_determined = _find_reasonable_pivot(
+                get_col(piv_col)[piv_row:], iszerofunc, simpfunc)
+
+            # _find_reasonable_pivot may have simplified some things
+            # in the process.  Let's not let them go to waste
+            for (offset, val) in newly_determined:
+                offset += piv_row
+                mat[offset*cols + piv_col] = val
```

Comment:
I believe this line of code can be omitted. This newly simplified entry is under the pivot position, so it will get zeroed when a multiple of the pivot row is added to the row containing this entry.